### PR TITLE
More unit tests for cluster and move some code out to make testing easier

### PR DESCRIPTION
**Pull request by asalkeld**
So testing Update() would be very hard as it pulls in a lot of code and you would have to mock a *lot* of calls.
I have tried to reduce the volume of calls by moving the kubenetes functions into a new package . This make the mocking you have to do in the cluster unit tests simpler and then we can unit test the kubeclient package easily too.

So my basic aim is to create another interface that we can mock, if there is a better way of splitting cluster up so we have more interfaces, I am all ears.

### CODE REVIEW

File: pkg/cluster/update.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/Azure/go-autorest/autorest/to"
 
 	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/kubeclient"
```
Comment (kargakis) [inline pkg/cluster/update.go]: Probably move it under pkg/util like the azure client?

File: pkg/cluster/update_test.go [hunk 2]
```diff
@@ -83,3 +98,426 @@ func TestFilterOldVMs(t *testing.T) {
...
+			gmc := gomock.NewController(t)
+			defer gmc.Finish()
+			kubeclient := mock_kubeclient.NewMockKubeclient(gmc)
+			virtualMachineScaleSetsClient := mock_azureclient.NewMockVirtualMachineScaleSetsClient(gmc)
+			virtualMachineScaleSetVMsClient := mock_azureclient.NewMockVirtualMachineScaleSetVMsClient(gmc)
+			mockListVMs(ctx, gmc, virtualMachineScaleSetVMsClient, tt.cs, "master", testRg, tt.vmsBefore["master"], nil)
+			mockListVMs(ctx, gmc, virtualMachineScaleSetVMsClient, tt.cs, "infra", testRg, tt.vmsBefore["infra"], nil)
+			mockListVMs(ctx, gmc, virtualMachineScaleSetVMsClient, tt.cs, "compute", testRg, tt.vmsBefore["compute"], nil)
+
+			if tt.expectDrain != nil {
```
Comment (kargakis) [inline pkg/cluster/update_test.go]: You shouldn't need this if statement, if it's nil, the loop will never run

File: pkg/cluster/update_test.go [hunk 3]
```diff
@@ -83,3 +98,426 @@ func TestFilterOldVMs(t *testing.T) {
...
+				ssc:        virtualMachineScaleSetsClient,
+				kubeclient: kubeclient,
+				log:        logrus.NewEntry(logrus.StandardLogger()),
+			}
+			got, err := u.getNodesAndDrain(ctx, tt.cs)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("simpleUpgrader.getNodesAndDrain() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
```
Comment (kargakis) [inline pkg/cluster/update_test.go]: Compare the actual error messages instead of expecting a random error

File: pkg/cluster/update_test.go [hunk 4]
```diff
@@ -83,3 +98,426 @@ func TestFilterOldVMs(t *testing.T) {
...
+				updateBlob:    updateBlob,
+				vmc:           virtualMachineScaleSetVMsClient,
+				ssc:           virtualMachineScaleSetsClient,
+				storageClient: storageClient,
+				kubeclient:    client,
+				log:           logrus.NewEntry(logrus.StandardLogger()).WithField("test", tt.name),
+			}
+			mockListVMs(ctx, gmc, virtualMachineScaleSetVMsClient, tt.cs, "master", testRg, tt.vmsList["master"], nil)
+
+			if err := u.waitForNewNodes(ctx, tt.cs, tt.nodes, tt.ssHashes); (err != nil) != tt.wantErr {
```
Comment (kargakis) [inline pkg/cluster/update_test.go]: compare error messages

File: pkg/util/kubeclient/ready.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+		return r, err
+	}
+
+	r, err = ready.PodIsReady(u.client.CoreV1().Pods("kube-system"), "master-api-"+computerName.toKubernetes())()
+	if !r || err != nil {
+		return r, err
+	}
+
+	r, err = ready.PodIsReady(u.client.CoreV1().Pods("kube-system"), "controllers-"+computerName.toKubernetes())()
+	if !r || err != nil {
```
Comment (kargakis) [inline pkg/util/kubeclient/ready.go]:
```go
return r, err
```
w/o the if statement

Comment (kargakis) [reply pkg/util/kubeclient/ready.go]: or `return ready.PodIsReady(...)`

File: pkg/util/kubeclient/ready_test.go [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+						},
+					},
+				},
+			}),
+		},
+	}
+	for _, tt := range tests {
+		u := &kubeclient{client: tt.kc}
+		got, err := u.MasterIsReady(tt.computerName)
+		if (err != nil) != tt.wantErr {
```
Comment (kargakis) [inline pkg/util/kubeclient/ready_test.go]: Compare error messages

File: pkg/util/kubeclient/ready_test.go [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+			for _, react := range tt.reactors {
+				tt.client.PrependReactor(react.verb, "nodes", react.reaction)
+			}
+			u := &kubeclient{
+				client: tt.client,
+				log:    logrus.NewEntry(logrus.StandardLogger()),
+			}
+
+			if err := u.WaitForReady(ctx, tt.role, tt.computerName); (err != nil) != tt.wantErr {
```
Comment (kargakis) [inline pkg/util/kubeclient/ready_test.go]: Compare error messages

File: pkg/cluster/update_test.go (hunk 3 above)
Comment (asalkeld) [reply pkg/cluster/update_test.go]: i don't have error testing here just yet, just the different logical paths..

**Comment by kargakis** [comment]
https://github.com/openshift/openshift-azure/issues/851

/retest

File: pkg/util/kubeclient/ready.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+var statefulsetWhitelist = []struct {
+	Name      string
+	Namespace string
+}{
+	{
+		Name:      "bootstrap-autoapprover",
+		Namespace: "openshift-infra",
+	},
+	{
+		Name:      "prometheus-forwarder",
```
Comment (mjudeikis) [inline pkg/util/kubeclient/ready.go]: This was removed, remove forwarder at all

Comment (asalkeld) [reply pkg/util/kubeclient/ready.go]: ok thanks, i'll compare the file to the latest

**Comment by kargakis** [comment]
https://github.com/openshift/openshift-azure/issues/804

hoping to see less of this flake with https://github.com/openshift/release/pull/2325

/retest

**Comment by jim-minter** [comment]
/hold cancel
/lgtm

this has been blocked too long on me - sorry.  This makes sense to me, I wonder whether a separate package under pkg/cluster would have been better than pkg/util/kubeclient, but it's good progress.

**Comment by jim-minter** [comment]
@kargakis I don't understand the message "Job ci/prow/etcdbackuprecovery has not succeeded." - what needs to be done to merge this?

**Pull request merged**